freestyle/shared/src/main/scala/freestyle/module.scala

```diff
@@ -16,171 +16,190 @@
-      }"""
-    c.Expr[Any](expr)
+      s"unable to find companion for $sym when expanding coproduct".invalidNel
+
+  private[this] def decodeAnnotationTarget(annottees: List[c.Expr[Any]]) =
+    annottees.map(_.tree) match {
+      case (cls: ClassDef) :: Nil =>
+        if (cls.mods.hasFlag(Flag.TRAIT | Flag.ABSTRACT)) cls.validNel
+        else
+          "The `@module` annotation can only be applied to a trait or an abstract class.".invalidNel
+
+      case (cls: ClassDef) :: more =>
+        "The trait or class annotated with `@module` must have no companion object.".invalidNel
+      case trees =>
+        "Unexpected trees $trees encountered for `@module` annotation".invalidNel
   }
 
-}
-
-object moduleImpl {
-
-  def impl(c: Context)(annottees: c.Expr[Any]*): c.universe.Tree = {
-    import c.universe._
-    import c.universe.internal.reificationSupport._
+  private[this] def gatherEffects(cls: ClassDef): ValidatedNel[String, List[(Name, Tree)]] =
+    cls.impl.body
+      .collect  { case valDef: ValDef if valDef.rhs.isEmpty => valDef }
+      .traverse { valDef => (valDef.name, valDef.tpt).validNel }
 
-    def fail(msg: String) = c.abort(c.enclosingPosition, msg)
+  private[this] def expandEffectType(tpe: Type): List[Type] = {
+    def isModuleFS(cs: ClassSymbol): Boolean =
+      cs.baseClasses.exists(_.name == TypeName("FreeModuleLike") )
 
-    def filterEffectVals(trees: Template): List[ValDef]  =
-      trees.collect { case v: ValDef if v.mods.hasFlag(Flag.DEFERRED) => v }
+    def expandClass(cs: ClassSymbol): List[Type] =
+      cs.info.decls.toList
+        .collect { case met: MethodSymbol if met.isAbstract => met }
+        .flatMap(x => expandEffectType(x.returnType))
 
-    val LL = freshTypeName("LL$")
-
-    def toImplArg(effVal: ValDef): ValDef = effVal match {
-      case q"$mods val $name: $eff[..$args]" =>
-        q"$mods val $name: $eff[$LL, ..$args]"
+    tpe.typeSymbol match {
+      case cs: ClassSymbol =>
+        if (isModuleFS(cs)) expandClass(cs) else List(tpe.typeConstructor)
+      case _ => Nil
     }
+  }
 
-    def mkModuleTrait(cls: ClassDef): ClassDef = {
-      val FF = freshTypeName("FF$")
-      // this is to make a TypeDef for `$FF[_]`
-      val wildcard = TypeDef(Modifiers(Flag.PARAM), typeNames.WILDCARD, List(), TypeBoundsTree(EmptyTree, EmptyTree))
-      val ffTParam = TypeDef(Modifiers(Flag.PARAM), FF, List(wildcard), TypeBoundsTree(EmptyTree, EmptyTree))
-
-      val ClassDef(mods, name, tparams, Template(parents, self, body)) = cls
-
-      val nbody = body.map {
-        case q"$mods val $name: $eff[..$args]" => q"$mods val $name: $eff[$FF, ..$args]"
-        case x => x
+  private[this] def typeToTypeTree(tpe: Type): ValidatedNel[String, Tree] =
+    Validated.catchNonFatal(c.parse(tpe.toString)).leftMap(t =>
+      NonEmptyList.of(s"unable to convert type $tpe to tree"))
+
+  private[this] lazy val defaultf0: ValidatedNel[String, Tree] =
+    q"""type Op[$AA] = Nothing""".validNel
```

Comment:
If it fits one line, single quotes is fine. 
Also, ¿why do you use the `Validated`? It seems pretty straightforward.